server/cluster/unsafe_recovery_controller.go

```diff
@@ -120,462 +182,754 @@ func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]st
+			zap.Uint64("store-id", storeID), zap.Uint64("expected-step", u.step), zap.Uint64("obtain-step", heartbeat.StoreReport.GetStep()))
+		// invalid store report, ignore
+		return false, nil
 	}
-	numFinished := 0
-	for _, peerReport := range report.PeerReports {
-		region := peerReport.RegionState.Region
-		if _, ok := toBeRemovedRegions[region.Id]; ok {
-			return false
-		} else if target, ok := targetRegions[region.Id]; ok {
-			if bytes.Equal(target.StartKey, region.StartKey) && bytes.Equal(target.EndKey, region.EndKey) && !u.containsFailedPeers(region) {
-				numFinished += 1
-			} else {
-				return false
+
+	if report, exists := u.storeReports[storeID]; exists {
+		// if receive duplicated report from the same TiKV, use the latest one
+		u.storeReports[storeID] = heartbeat.StoreReport
+		if report == nil {
+			u.numStoresReported++
+			if u.numStoresReported == len(u.storeReports) {
+				return true, nil
 			}
 		}
 	}
-	return numFinished == len(targetRegions)
+	return false, nil
 }
 
-type regionItem struct {
-	region *metapb.Region
+// Gets the stage of the current unsafe recovery.
+func (u *unsafeRecoveryController) GetStage() unsafeRecoveryStage {
+	u.RLock()
+	defer u.RUnlock()
+	return u.stage
 }
 
-func (r regionItem) Less(other btree.Item) bool {
-	return bytes.Compare(r.region.StartKey, other.(regionItem).region.StartKey) < 0
+func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
+	u.stage = stage
+
+	var output StageOutput
+	output.Time = time.Now().Format("2006-01-02 15:04:05.000")
+	switch u.stage {
+	case idle:
+	case collectReport:
+		stores := ""
+		count := 0
+		for store := range u.failedStores {
+			count += 1
+			stores += fmt.Sprintf("%d", store)
+			if count != len(u.failedStores) {
+				stores += ", "
+			}
+		}
+		output.Info = fmt.Sprintf("Unsafe recovery enters collect report stage: failed stores %s", stores)
+	case forceLeaderForCommitMerge:
+		output.Info = "Unsafe recovery enters force leader for commit merge stage"
+		output.Actions = u.getForceLeaderPlanDigest()
+	case forceLeader:
+		output.Info = "Unsafe recovery enters force leader stage"
+		output.Actions = u.getForceLeaderPlanDigest()
+	case demoteFailedVoter:
+		output.Info = "Unsafe recovery enters demote failed voter stage"
+		output.Actions = u.getDemoteFailedVoterPlanDigest()
+	case createEmptyRegion:
+		output.Info = "Unsafe recovery enters create empty region stage"
+		output.Actions = u.getCreateEmptyRegionPlanDigest()
+	case finished:
+		u.cluster.PauseOrResumeScheduler("all", 0)
+		if u.step > 1 {
+			// == 1 means no operation has done, no need to invalid cache
+			u.cluster.DropCacheAllRegion()
+		}
+		output.Info = "Unsafe recovery finished"
+	case failed:
+		u.cluster.PauseOrResumeScheduler("all", 0)
+		output.Info = fmt.Sprintf("Unsafe recovery failed: %v", u.err)
+	}
+
+	u.output = append(u.output, output)
+	data, err := json.Marshal(output)
+	if err != nil {
+		panic(err)
```

Comment:
Should we panic? Maybe an error log is enough.

Reply:
it can't happen, so I just panic straightforward here

Reply:
Yes, but this is only a log problem, we don't need to panic here.